src/realm/obj.cpp

```diff
@@ -1103,26 +1137,31 @@ void Obj::to_json(std::ostream& out, size_t link_depth, const std::map<std::stri
                 }
                 else if (it.second.is_type(type_TypedLink)) {
                     auto obj_link = it.second.get<ObjLink>();
+                    auto obj_key = obj_link.get_obj_key();
                     auto target_table = m_table->get_parent_group()->get_table(obj_link.get_table_key());
+                    auto primary_key_coll = target_table->get_primary_key_column();
 
-                    if (output_mode == output_mode_xjson && !target_table->is_embedded() &&
-                        !obj_link.is_unresolved()) {
-                        out_mixed_xjson(out, obj_link.get_obj_key());
+                    if (target_table->is_embedded()) {
+                        size_t new_depth = link_depth == not_found ? not_found : link_depth - 1;
+                        auto link = it.second.get<ObjKey>();
+                        target_table->get_object(link).to_json(out, new_depth, renames, followed, output_mode);
                     }
-                    else if (output_mode == output_mode_xjson_plus && !target_table->is_embedded() &&
-                             !obj_link.is_unresolved()) {
-                        out << "{ \"$link\": { \"table\": \"" << obj_link.get_table_key() << "\", \"key\": ";
-                        out_mixed_xjson(out, obj_link.get_obj_key());
-                        out << "}}";
+                    else if (primary_key_coll && output_mode == output_mode_xjson) {
+                        out_mixed_xjson(out, target_table->get_primary_key(obj_key));
```

Comment:
I guess that the name of the target table should be output around here?

Reply:
The table name is outputted above with the $embeddedDictionary / $linkedDictionary annotation. This is not an official spec though (hence xjson_plus being the output mode), but I figured that we may as well only decode the table name once for a dictionary of embedded / linked object

Reply:
Ah - I see it now. We should just be aware that at some point in the future, a TypedLink can point to any table.